### DISCUSSION TITLE
Probelm running POC mode without GPU

### DISCUSSION THREAD

**Discussion by falibabaei** [NVIDIA/NVFlare#1852] [General]
Hi,
I have the following error when I try to start poc mod
nvflare poc --start -gpu 0
Failed to get host gpu ids:('Failed to call nvidia-smi with query index', '')
how can I disable the usage of GPU?

**Comment by chesterxgchen** [comment]
if you don't have gpu, you can simple use 
nvflare poc --start

**Reply by falibabaei** [reply]
I use that but still fail to start and the same error occurs

**Reply by chesterxgchen** [reply]
are you installed via pip install ? if yes, which version of nvflare are you using ? 

are you directly using source code (github clone) ? if yes, which branch are you using

**Reply by chesterxgchen** [reply]
you can find out the version number by 
```
nvflare --version
```

**Reply by SYangster** [reply]
can you also try running `nvidia-smi --query-gpu=index --format=csv` and see what the returncode is?

**Reply by falibabaei** [reply]
Hi,
thank you very much for your answer. Yes, I installed with pip and I clone the source code from github.
 git branch --show-curren
dev
nvflare --version
NVFlare version is 2.3.1
nvidia-smi --query-gpu=index --format=csv
NVIDIA-SMI has failed because it couldn't communicate with the NVIDIA driver. Make sure that the latest NVIDIA driver is installed and running.
Is it because I cloned the dev branch? also, I have an external GPU. If it is not connected, I can not start the poc mod.

**Reply by falibabaei** [reply]
The result when I connect the GPU
nvidia-smi --query-gpu=index --format=csv
index
0

**Reply by SYangster** [reply]
That result of "index 0" does mean that the GPU is detected when connected which is good. Are you able to run the POC successfully with the external GPU connected?

Also the dev branch shouldn't be the issue, however you can check by switching to the stable 2.3 branch (`git checkout 2.3`) just to be sure.

Currently we are checking for the presence of a GPU using `nvidia-smi --query-gpu=index --format=csv` and it seems that when your external GPU is not connected, the query fails with the "couldn't communicate with the NVIDIA driver" rather than just returning that there is no GPU. Maybe you could also try `nvidia-smi -L` when the GPU is not connected, which might be a better way for us to check?

**Reply by falibabaei** [reply]
Yes it works with the external GPU. I will try and let you know

**Reply by falibabaei** [reply]
Hi, 
I switched to branch 2.3, but the same error occurred. The results of the command are as follows:
git branch --show-current
2.3
nvflare poc --start
Failed to get host gpu ids:('Failed to call nvidia-smi with query index', '')
nvidia-smi -L
NVIDIA-SMI has failed because it couldn't communicate with the NVIDIA driver. Make sure that the latest NVIDIA driver is installed and running.

**Reply by falibabaei** [reply]
I installed version =2.4.0rc1 and it works without GPU

**Reply by chesterxgchen** [reply]
Great. I must have relax this requirement in 2.4.0RC1  when I made the change :-) . I haven't update the documentation with 2.4.0rc1 POC changes. POC in RC1 is "Local, Production mode, non-HA", so it will be working slightly different from  2.3.x POC, for one, since it's production mode, the FLARE API session must be in secure session ( like in production) ( prev. POC is non-secure session). Also the admin user is not "admin", but "admin@nvidia.com".  I think that's all the subtle difference. It has more functionality. rest should behave the same

**Reply by falibabaei** [reply]
Hi,
As you mentioned, POC in RC1 is "Local, Production mode, non- HA". I am preparing some slides to present to my team. Before this release you had a warning as follows
POC mode is not intended to be secure and should not be run in a production environment or any environment where the server's ports are exposed.
Should I remove this warning in the new version?

**Reply by nvkevlu** [reply]
Yes, I am working on updating the documentation to reflect the updates for POC mode, and as part of these updates, the warnings are going to be removed.

**Reply by chesterxgchen** [reply]
even in that case, POC in 2.4.0 will be "local", which means both server and clients will be on the same machine using localhost. 
in real production, this is never the case.

**Reply by chesterxgchen** [reply]
do you mind share which company and team do you work with ?

**Reply by falibabaei** [reply]
I am working at Karlsruhe Institute of Technology (KIT), Germany, as a data analysis

**Reply by chesterxgchen** [reply]
Thank you ! We have several Germany customers.

**Reply by falibabaei** [reply]
We are still exploring different frameworks. Most of our users are researchers and as a researcher  I can say nvflare is not the easiest framework to work with (just a feedback😀). But for the production I think it is the best one.

**Reply by chesterxgchen** [reply]
Understand. We are trying to improve and make it easy to get on going without learning NVFLARE concepts (like you need to right now).
Once after 2.4.0 release, this should dramatically changed. for example. Look at these new examples (still need polish and documentation): 

https://github.com/NVIDIA/NVFlare/tree/dev/examples/advanced/ml-to-fl

In these two examples (one with client API, another one with Decorator approach, we don't need to learn anything about NVFLARE to get started, just few lines code changes in few places: receive_model, submit_model, submit_metrics, send_model. The rest will take care by NVFLARE framework. At the moment, you need to know the configurations, we are working to make that part simpler as well, expose less NVFLARE concept, or auto generate initial config for you, you can later modify to make it more complicated once you learn more).  Of course, you will need to learn more as time goes one and you are move to production. 

In this PR: https://github.com/NVIDIA/NVFlare/pull/1867 we directly integrated with pytorch-lightning. Similarly, with just few lines of code changes (without changing the existing code structure), you are turning DL code into NVFLARE Fed Learning Code. 

We hope with these changes, getting started with with NVFLARE is really simple, especially for researcher trying to run FL applications. 

Once you changed the code, you can start the POC model with 
```
nvflare poc --prepare -ex admin@nvidia.com
```
you have the 2 clients and 1 server started (without admin console). 

Then we will add new CLI command ( still in design) ( something like)

```
nvflare job submit --job_folder <folder>  -f <some kind of config key=value pairs>
```
directly submit job from CLI without using Admin Console. (leverage flare API)

Of course, if you are developing FL algorithms, you might need to learn more. NVFLARE Is component-based, event-driven so everything is customizable. To know how to customize you will need to understand how it put together, what are the events fired etc.

**Reply by falibabaei** [reply]
That is perfect. Thank you very much for sharing with me. I am excited to see the new changes

**Reply by chesterxgchen** [reply]
> ...I can say nvflare is not the easiest framework to work with (just a feedback😀)

We would appreciate if you can share some detailed feedbacks
* what doesn't work for you
* what is not easy
* what you wish to see
* etc.

**Reply by falibabaei** [reply]
I will create a list and share it with you. Additionally, I plan to present the NVFlare to my team next week and gather their feedback.

**Reply by chesterxgchen** [reply]
that would be great. You can directly email me chesterc@nvidia.com

**Reply by falibabaei** [reply]
Sure

**Reply by falibabaei** [reply]
Hi,
I have 
nvflare --version
NVFlare version is 2.4.0rc1
 and I am trying to submit a job from the [examples](https://github.com/NVIDIA/NVFlare/tree/dev/examples/hello-world/ml-to-fl) you sent me.
I have the error
Error processing ['/home/se1131/NVFlare/examples/hello-world/hello-pt/workspace/site-2/startup/../4d3dff43-9775-43f9-a45b-eeb055c77a02/app_site-2/config/config_fed_client.json'] in JSON element {'name': 'PTFilePipeLauncherExecutor', 'args': {'launcher_id': 'launcher', 'heartbeat_timeout': 60}}: path: executors.#1.executor, exception: Cannot find component class "PTFilePipeLauncherExecutor"

**Reply by chesterxgchen** [reply]
The examples are merged just last week, they are not in 2.4.0rc1. 

You can try this on dev branch ( bleeding edge, unstable, development branch). 
assuming ```/home/se1131/NVFlare/```  is your git clone directory for NVFlare. 

python -m pip install -e /home/se1131/NVFlare/[dev]

or 
```
cd /home/se1131/NVFlare/

python -m pip install -e .[dev]

```

then you can run the example.

**Reply by falibabaei** [reply]
Sorry, but it still encourages the same error

     raise ConfigError("Config error in {}: {}".format(self.config_file_name, secure_format_exception(e))) nvflare.fuel.common.excepts.ConfigError: Config error in ['/tmp/nvflare/poc/example_project/prod_00/site-1/startup/../ae216c7a-ad56-4649-8ac2-7e740884d87f/app_site-1/config/config_fed_client.json']: Error processing ['/tmp/nvflare/poc/example_project/prod_00/site-1/startup/../ae216c7a-ad56-4649-8ac2-7e740884d87f/app_site-1/config/config_fed_client.json'] in JSON element {'name': 'PTFilePipeLauncherExecutor', 'args': {'launcher_id': 'launcher', 'heartbeat_timeout': 60}}: path: executors.#1.executor, exception: Cannot find component class "PTFilePipeLauncherExecutor"

**Reply by chesterxgchen** [reply]
may sure you are using the correct nvflare version. 
assume you have done the git pull to get the latest dev branch. Can you find the class "PTFilePipeLauncherExecutor" in the source code ? 

@YuanTingHsieh can you help ?

**Reply by falibabaei** [reply]
here is my nvflare version
nvflare --version
NVFlare version is 2.4.0.dev+30.gb2a46a61.dirty
But I can not find PTFilePipeLauncherExecutor

**Reply by chesterxgchen** [reply]
it should be here : https://github.com/NVIDIA/NVFlare/blob/dev/nvflare/app_opt/pt/file_pipe_launcher_executor.py

**Reply by falibabaei** [reply]
I changed the version and it worked. Thanks
